src/persist-client/src/batch.rs

```diff
@@ -328,47 +359,126 @@ where
-            self.parts.write(part, upper, since).await;
+        // if we've filled up a batch part, flush out to blob to keep our memory usage capped.
+        if size + self.current_part_total_bytes > self.blob_target_size {
+            // TODO: we currently consolidate a part when we've reached our target size of
+            // _unconsolidated_ updates. This means the part we actually write to Blob may
+            // be (substantially) smaller post-consolidation. We could alternatively create
+            // a `BTreeMap<(&[u8], &[u8], T), D>`, consolidate as we go, and only write out
+            // a part once the data referenced by the map reaches our target size.
+            self.flush_current_part().await;
             part_written = true;
         }
 
+        self.current_part_total_bytes += size;
+        self.current_part_key_bytes += k_range.len();
+        self.current_part_value_bytes += v_range.len();
+
         if part_written {
+            // NB: we flush our shared buf's contents to Blob storage when we've hit our
+            // target size, but we can only determine whether we've hit the target size
+            // after encoding into the buf. this means that after flushing to Blob, we have
+            // one dangling update that we need to copy back after the buf is reset.
+            let dangling_key = self.key_buf[k_range].to_owned();
+            let dangling_val = self.val_buf[v_range].to_owned();
+
+            self.key_buf.clear();
+            self.val_buf.clear();
+
+            self.key_buf.extend_from_slice(&dangling_key);
+            self.val_buf.extend_from_slice(&dangling_val);
+
+            self.current_part
+                .push(((0..dangling_key.len(), 0..dangling_val.len()), ts, diff));
             Ok(Added::RecordAndParts)
         } else {
+            self.current_part.push(((k_range, v_range), ts, diff));
             Ok(Added::Record)
         }
     }
+
+    /// Flushes the current part to Blob storage, first consolidating and then columnar encoding
+    /// the updates. It is the caller's responsibility to chunk `current_part` to be no greater
+    /// than [crate::PersistConfig::blob_target_size], and must absolutely be less than
+    /// [mz_persist::indexed::columnar::KEY_VAL_DATA_MAX_LEN]
+    async fn flush_current_part(&mut self) {
+        let mut updates = Vec::with_capacity(self.current_part.len());
+        for ((k_range, v_range), t, d) in self.current_part.drain(..) {
+            updates.push(((&self.key_buf[k_range], &self.val_buf[v_range]), t, d));
+        }
+
+        let start = Instant::now();
+        consolidate_updates(&mut updates);
+        self.batch_write_metrics
+            .step_consolidation
+            .inc_by(start.elapsed().as_secs_f64());
+
+        if updates.is_empty() {
+            return;
+        }
+
+        self.num_updates += updates.len();
+
+        match (&self.greatest_kv_seen, updates.last()) {
```

Comment:
it took me quite a while to convince myself how this worked. maybe if we pull out an outer if let for greatest_kv_in_part?

```
if let Some(greatest_kv_in_part) = greatest_kv_in_part {
    match self.greatest_kv_seen.as_ref() {
        ...
    }
}
```

Reply:
given the bug fixes, this whole section has been very rewritten and I think reads a lot more easily than before. lmk what you think